File: misc/aooswd/ex/bool_ex_interpreter/java/NotExp.java
/**
 * @author dgayler
 *
 */
public class NotExp implements BooleanExp
{

	private BooleanExp child;
	
	public NotExp (BooleanExp child)
	{
		this.child = child;
	}
	
	/**
	 * @see BooleanExp#Evaluate(Context)
	 */
	public boolean evaluate(Context con)
	{
		return !child.evaluate(con);
	}

	/**
	 * @see BooleanExp#generateCode(Context)
	 */
	public String generateCode(Context con)
	{
		String childName = child.generateCode (con);
		con.addCodeStatement(new LoadStatement (childName));
		con.addCodeStatement (new NotStatement());
		String temp = con.getTemp();
		con.addCodeStatement (new StoreStatement (temp));
		return temp;
	}

}

File: misc/aooswd/ex/bool_ex_interpreter/cpp/not_exp.cpp
#include "not_exp.h"
#include "load_statement.h"
#include "not_statement.h"
#include "store_statement.h"
#include "context.h"

string NotExp::generateCode (Context& c)
{
  string childName = child -> generateCode (c);
  c.addCodeStatement (new LoadStatement (childName));
  c.addCodeStatement (new NotStatement ());
  string temp = c.getTemp();
  c.addCodeStatement (new StoreStatement (temp));
  return temp;
}

